src/devices/somfy.c

```diff
@@ -0,0 +1,193 @@
+
+Retransmission frames' preamble:
+
+    °°°°____°°°°____°°°°____°°°°____°°°°____°°°°____°°°°____°°°°°°°°_
+
+During reception, I observed that for both preambles the last low value is sometimes missing. Hence, I just call the
+manchester decoder a second time with one bit offset, if the first decoding failed.
+
+The data is manchester encoded _° represents a 1 and °_ represents a 0. The data section consists of 56 bits that equals
+7 bytes of scrambled data. The data is scrambled by XORing each following byte with the last scrambled byte. After
+descrambling, the 7 bytes have the following meaning conting byte from left to right as in big endian byte order:
+
+- byte 0:   called "random" in [1] and "key" in [2], in the end it is just the seed for the scrambler
+- byte 1:   The higher nibble represents the control command, the lower nibble is the frame's checksum calculated by XORing 
+            all nibbles
+- byte 2-3: Replay counter value in big endian byte order
+- byte 4-6: Remote control channel's address
+*/
+
+#include "decoder.h"
+
+static const char *control_str[] = {
+    "? (0)",
+    "My (1)",
+    "Up (2)",
+    "My + Up (3)",
+    "Down (4)",
+    "My + Down (5)",
+    "Up + Down (6)",
+    "? (7)",
+    "Prog (8)",
+    "Sun + Flag (9)",
+    "Flag (10)",
+    "? (11)",
+    "? (12)",
+    "? (13)",
+    "? (14)",
+    "? (15)",
+};
+
+static int somfy_rts_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    data_t *data;
+
+    uint8_t is_retransmission = 0;
+    uint8_t decode_row = 0;
+    uint8_t data_start = 0;
+    const uint8_t *preamble_pattern;
+    unsigned preamble_pattern_bit_length = 0;
+
+    bitbuffer_t bitbuffer_decoded = { 0 };
+
+    uint8_t message_bytes[7] = { 0 };
+    uint8_t chksum_calc = 0;
+    uint8_t chksum = 0;
+    uint16_t counter = 0;
+    char address[7];
+    uint8_t control = 0;
+    uint8_t seed = 0;
+
+    for (int i = 0; i < bitbuffer->num_rows; i++) {
+        if (bitbuffer->bits_per_row[i] > 170) {
+            is_retransmission = 1;
+            decode_row = i;
+            data_start = 65;
+            preamble_pattern = (const uint8_t *) "\xf0\xf0\xf0\xf0\xf0\xf0\xf0\xff";
+            preamble_pattern_bit_length = 64;
+            break;
+        } else if (bitbuffer->bits_per_row[i] > 130) {
+            is_retransmission = 0;
+            decode_row = i;
+            data_start = 25;
+            preamble_pattern = (const uint8_t *) "\xf0\xf0\xff";
+            preamble_pattern_bit_length = 24;
+            break;
+        }
+    }
+
+    if (data_start == 0)
+        return DECODE_FAIL_SANITY;
+
+    if (bitbuffer_search(bitbuffer, decode_row, 0, preamble_pattern, preamble_pattern_bit_length) != 0)
+        return DECODE_FAIL_SANITY;
+
+    if (bitbuffer_manchester_decode(bitbuffer, decode_row, data_start, &bitbuffer_decoded, 56) - data_start < 56)
+        if (bitbuffer_manchester_decode(bitbuffer, decode_row, data_start - 1, &bitbuffer_decoded, 56) - data_start + 1 < 56)
```

Comment:
if you consider the `bitbuffer_search` value you don't need to try `-1` here?

Reply:
I could also write: `if (bitbuffer_manchester_decode(bitbuffer, decode_row, data_start - 1, &bitbuffer_decoded, 56) - (data_start - 1) < 56)`.
Is it what you mean?

Reply:
I meant `data_start = bitbuffer_search()` ;)

Reply:
Sorry, it seems like the wrong code portion was displays here. I expect `bitbuffer_search` to return 0 if the preamble is found at the beginning of the bitbuffer. How can `bitbuffer_search` return `-1`?

Reply:
Sorry, I see. The preamble is 24 bits, but you also try to decode with one bit ignored (.i.e +1) after the preamble. Are there both types of transmissions ("last low bit ok/missing" in your rtl_433_tests PR? I'd like to take a peek :)

Reply:
I added it now.

Reply:
With the fix from https://github.com/merbanan/rtl_433/commit/97fdf16333a5f1a7fb5490996a942186be5cf99e my workaround is not needed anymore. Hence, I removed the second call to the Manchester decoder.

Reply:
The preamble should then likely be 0xe1e1fe (one 0-bit added) and bitbuffer_search() will yield the proper data_start. Or perhaps use 0xf0f0ff00 with a preamble_len of 25.